eclair-core/src/main/scala/fr/acinq/eclair/router/Router.scala

```diff
@@ -573,26 +574,58 @@ class Router(val nodeParams: NodeParams, watcher: ActorRef, initialized: Option[
-              case None => log.warning("received query for shortChannelId={} that we don't have", shortChannelId)
-              case Some(ca) =>
-                if (QueryShortChannelIdsTlv.QueryFlagType.includeAnnouncement(flag)) {
-                  transport ! ca
-                  c1 = c1 + 1
-                }
-                if (QueryShortChannelIdsTlv.QueryFlagType.includeUpdate1(flag)) d.updates.get(ChannelDesc(ca.shortChannelId, ca.nodeId1, ca.nodeId2)).foreach { u => transport ! u; u1 = u1 + 1 }
-                if (QueryShortChannelIdsTlv.QueryFlagType.includeUpdate2(flag)) d.updates.get(ChannelDesc(ca.shortChannelId, ca.nodeId2, ca.nodeId1)).foreach { u => transport ! u; u1 = u1 + 1 }
+      val flags = routingMessage.queryFlags_opt.map(_.array).getOrElse(List.empty[Long])
+
+      // we loop over channel ids and query flag and send what the sender requested
+      // we keep track of how many announcements and updates we've sent, and we also track node Ids for node announcement
+      // we've already sent to avoid sending them multiple times, as requested by the BOLTs
+      @tailrec
+      def loop(ids: List[ShortChannelId], flags: List[Long], numca: Int = 0, numcu: Int = 0, sent: Set[PublicKey] = Set.empty[PublicKey]): (Int, Int, Int) = ids match {
+        case Nil => (numca, numcu, sent.size)
+        case head :: tail if !d.channels.contains(head) =>
+          log.warning("received query for shortChannelId={} that we don't have", head)
+          loop(tail, flags.drop(1), numca, numcu, sent)
+        case head :: tail =>
+          var numca1 = numca
+          var numcu1 = numcu
+          var sent1 = sent
+          val ca = d.channels(head)
+          val flag_opt = flags.headOption
+          // no flag means send everything
+
+          if (flag_opt.map(QueryShortChannelIdsTlv.QueryFlagType.includeChannelAnnouncement).getOrElse(true)) {
+            transport ! ca
+            numca1 = numca1 + 1
+          }
+          if (flag_opt.map(QueryShortChannelIdsTlv.QueryFlagType.includeUpdate1).getOrElse(true)) {
+            d.updates.get(ChannelDesc(ca.shortChannelId, ca.nodeId1, ca.nodeId2)).foreach { u =>
+              transport ! u
+              numcu1 = numcu1 + 1
             }
-            (c1, u1)
-        }
-      log.info("received query_short_channel_ids with {} items, sent back {} channels and {} updates", shortChannelIds.array.size, channelCount, updatesCount)
+          }
+          if (flag_opt.map(QueryShortChannelIdsTlv.QueryFlagType.includeUpdate2).getOrElse(true)) {
+            d.updates.get(ChannelDesc(ca.shortChannelId, ca.nodeId2, ca.nodeId1)).foreach { u =>
+              transport ! u
+              numcu1 = numcu1 + 1
+            }
+          }
+          if (flag_opt.map(QueryShortChannelIdsTlv.QueryFlagType.includeNodeAnnouncement1).getOrElse(true) && !sent1.contains(ca.nodeId1)) {
+            d.nodes.get(ca.nodeId1).foreach { a =>
+              transport ! a
```

Comment:
Our convention is `c` for channels, `u` for updates and `n` for nodes.

```suggestion
              transport ! n
```